환경설정 파일 변경 및 소스코드 업로드

seyoon-kim opened the pull request:
**데모페이지**
[데모페이지 이동](http://10.78.65.35:8000/examples/)

**소개**
- Selector API의 구현인 querySelector와 querySelectorAll를 직접 구현한다.
- 실제 구현 전에 요구사항에 맞는 TC를 우선 작성 후 공유한다.

**요구사항**
- getElementById, getElementByTagName을 이용해 구현한다. (특정 IE에서 이슈가 있는 getElementByClassName도 직접 구현한다.)
- .class, #id, tag 3개의 셀렉터만 이용할 수 있다.
- 셀렉터의 조합으로 엘리먼트를 찾을 수 있다 배열로 엘리먼트를 직접 반환함.

```
domutil.querySelectorAll('.class');
domutil.querySelector('#id');

//myClass라는 클래스를 가진 엘리먼트의 자식요소중 div를 모두 리턴
domutil.querySelectorAll('.myClass div');

```

seyoon-kim committed:
환경설정 파일 변경 및 소스코드 업로드

seyoon-kim committed:
Update, Demo page작성 및 소스정리

- Demo page작성
- 소스코드 정리

dongsik-yoo commented:
버그가 있는거 같네요. 중복으로 나오는 경우가 있어요.
```js
Domutil.querySelectorAll('div li');
Domutil.querySelectorAll('div ol');
```
셀렉터가 요구 사항에 맞지 않는 것일까요?

dongsik-yoo commented:
스크립트 오류가 납니다.

```js
Domutil.querySelector('.area #date');

index.js:42 Uncaught TypeError: rootEle.getElementById is not a function
    at _findElementsOfMatchingSelector (index.js:42)
    at _makeArrayMatchingToSelctor (index.js:76)
    at Object.querySelector (index.js:89)
    at <anonymous>:1:9
```

dongsik-yoo commented:
찾지 못한 경우에 빈 배열이 나와야 할 것 같아요. 
```js
Domutil.querySelectorAll('#titl');
```
![image](https://user-images.githubusercontent.com/26706716/30358122-862fe5b0-987d-11e7-8c5f-e792b9c2a63e.png)

dongsik-yoo commented on package.json:
```diff
@@ -16,6 +17,7 @@
     "karma-chrome-launcher": "^2.2.0",
     "karma-jasmine": "^1.1.0",
     "karma-webpack": "^2.0.4",
+    "local-web-server": "^2.2.4",
```
`local-web-server`는 전역으로 설치해야 동작하는 것 같으니 package.json 에서는 빼도 될 것 같아요.

seyoon-kim replied:
네 해당 패키지 제외하였습니다.

dongsik-yoo commented on test/spec.js:
```diff
@@ -0,0 +1,92 @@
+    });
+});
+
+xdescribe('getElementsByClassName Pollyfil', function() {
```
여기는 테스트 안 하는 이유가 있나요? `xdescribe` 처리 되어 있네요

seyoon-kim replied:
원래 getElementsByClassNamePollyfil 함수를 전역에 두고 사용했습니다. 그러나 현재는 Domutil 내부에서만 사용한다고 판단되어 이전에 전역에서 테스트 하던것을 사용하지 않게되어 xdescribe해 두었습니다.

dongsik-yoo replied:
테스트가 필요한 부분이면, 테스트 목적으로 외부에 노출시켜서 테스트해도 됩니다.

seyoon-kim replied:
넵, 해당하는 함수만 describe안에 넣어두어 테스트 할 수 있도록 변경하였습니다.

dongsik-yoo commented on src/index.js:
```diff
@@ -0,0 +1,119 @@
+     */
+    var _findElementsOfMatchingSelector = function(rootEle, selectors) {
+        var result;
+        var rClasName = /^\./g;
```
오타인듯^^ ㅎㅎ rClassName

seyoon-kim replied:
넵 rClassName으로 수정하였습니다 ^^;

dongsik-yoo commented on src/index.js:
```diff
@@ -0,0 +1,119 @@
+        if (rClasName.test(selectors)) {
+            selectors = selectors.replace(rClasName, '');
+            result = _getElementsByClassNamePolyfill(rootEle, selectors);
+            result = Array.prototype.slice.call(result);
```
IE8에서 사용할 수 있는지 한 번 확인해 보세요. 이 경우는 `result`가 확실히 배열이므로 쓸 수가 있겠네요.

seyoon-kim replied:
네 IE8 테스트 결과 result 변수값을 초기화 하지않으니 에러가 발생하였습니다. result 변수에 []로 초기화 하였더니 문제가 해결되었습니다.

dongsik-yoo commented on src/index.js:
```diff
@@ -0,0 +1,119 @@
+    // getElementsByClassName*() IE 8 이하 버전 호환
+    var _getElementsByClassNamePolyfill = function(rootEle, selector) {
+        var allElements;
+        var result;
```
`result` 변수의 초기화 코드가 보이지 않습니다.

dongsik-yoo replied:
데모페이지의 실제 코드에는 초기화 되어 있네요

seyoon-kim replied:
네 result 변수 배열로 초기화하였습니다.

seyoon-kim committed:
Update, _findElementsOfmatchingSelector 함수 수정

- 셀렉터가 ID인 경우 getElementById()함수를 사용해야 합니다. 그런데 앞에 document가 아니라 rootEle라는 전에 검색된 요소의 배열로 요청되어 함수가 실행이 안되었습니다. 그래서 rootEle -> document 수정하였습니다.

dongsik-yoo commented on src/index.js:
```diff
@@ -0,0 +1,119 @@
+
+    var querySelector = function(selectors) {
+        var from = [];
+        var result = [];
```
querySelector는 리턴이 배열이 아니지 않나요?

seyoon-kim replied:
네 원래 querySelector는  요소를 반환하지만 현재 과제의 요구사항에는 "배열로 엘리멘트를 직접 반환함"으로 나와있어서 배열로 반환하게 만들었습니다.

dongsik-yoo commented:
1차 리뷰는 완료하였습니다.
내일 11시 30분에 리뷰 시간 된다고들 하셨으니 일정으로 생성해서 공유해 주세요.

seyoon-kim committed:
Update, 셀렉터가 id인 경우 해당하는 요소를 찾지못한 경우

셀렉터가 id인 경우에 해당하는 요소를 찾지 못한 경우 빈배열을 반환하도록 수정하였습니다.

seyoon-kim committed:
Update, 리뷰사항 수정

1. pakage.json 로컬웹서버 글로벌설치로 인한 local-web-sever 삭제
2. spec.js _getElementsByClassNamePolyfill 테스트케이스 안에 함수 생성하여 테스트
3. rClasName, rClassName 오타수정
4. result 변수 배열로 초기화

seyoon-kim committed:
Update, querySelectorAll의 반환값 배열 요소중 중복되는 요소 제거

querySelectorAll의 반환값의 배열 요소가 중복도는 요소가 발견되었습니다. 이를 해결 하기 위해 removeSameElement라는 함수를 만들어 중복되는 값을 제거하도록 하였습니다.
